python/ray/autoscaler/v2/instance_manager/cloud_providers/kuberay/cloud_provider.py

```diff
@@ -0,0 +1,536 @@
+        if request_id in self._requests:
+            # This request is already processed.
+            return
+        self._requests.add(request_id)
+
+        scale_request = self._initialize_scale_request(
+            to_launch=shape, to_delete_instances=[]
+        )
+
+        if scale_request.worker_groups_with_pending_deletes:
+            error_msg = (
+                "There are workers to be deleted from: "
+                f"{scale_request.worker_groups_with_pending_deletes}. "
+                "Waiting for them to be deleted before creating new workers."
+            )
+            logger.warning(error_msg)
+            self._add_launch_errors(
+                shape,
+                request_id,
+                details=error_msg,
+            )
+            return
+
+        try:
+            self._submit_scale_request(scale_request)
+        except Exception as e:
+            logger.exception(f"Error launching nodes: {e}")
+            self._add_launch_errors(shape, request_id, str(e), e)
+
+    def poll_errors(self) -> List[CloudInstanceProviderError]:
+        errors = []
+        errors += self._launch_errors_queue
+        errors += self._terminate_errors_queue
+        self._launch_errors_queue = []
+        self._terminate_errors_queue = []
+        return errors
+
+    ############################
+    # Private
+    ############################
+
+    def _initialize_scale_request(
+        self, to_launch: Dict[NodeType, int], to_delete_instances: List[CloudInstanceId]
+    ) -> "KuberayProvider.ScaleRequest":
+        """
+        Initialize the scale request based on the current state of the cluster and
+        the desired state (to launch, to delete).
+
+        Args:
+            to_launch: The desired number of workers to launch for each node type.
+            to_delete_instances: The instances to delete.
+
+        Returns:
+            The scale request.
+        """
+
+        # Update the cached states.
+        self._sync_with_api_server()
+        ray_cluster = self.ray_cluster
+        cur_instances = self.instances
+
+        # Get the worker groups that have pending deletes and the worker groups that
+        # have finished deletes.
+        (
+            worker_groups_with_pending_deletes,
+            worker_groups_without_pending_deletes,
+        ) = self._get_workers_groups_with_deletes(
+            ray_cluster, set(cur_instances.keys())
+        )
+
+        # Calculate the desired number of workers by type.
+        num_workers_dict = defaultdict(int)
+        for instance_id, instance in cur_instances.items():
+            if instance.node_kind == NodeKind.HEAD:
+                # Only track workers.
+                continue
+            num_workers_dict[instance.node_type] += 1
+
+        # Add to launch nodes..
+        for node_type, count in to_launch.items():
+            num_workers_dict[node_type] += count
+
+        # Remove deleting workers.
```

Comment:
The two loops `for to_delete_id in to_delete_instances` and `for instance_id in to_delete_instances` are very similar. Can we fuse them? If we can't, we should at least have a consistent naming convention (`to_delete_id`, `instance_id`).